shell/common/pipeline.h

```diff
@@ -36,6 +36,20 @@ size_t GetNextPipelineTraceID();
 
 /// A thread-safe queue of resources for a single consumer and a single
 /// producer.
+///
+/// Pipelines support two key operations: produce and consume.
+///
+/// The consumer calls `Consume` to wait for a resource to be produced and
+/// consume it when ready.
+///
+/// The producer calls `Produce` to generate a `ProducerContinuation` which
+/// provides a means to enqueue a resource in the pipeline. When the resource
+/// has been prepared, the producer calls `Complete` on the continuation, which
+/// enqueues the resource and signals the waiting consumer.
+///
```

Comment:
Would it make sense to document this on the methods themselves as well?

Reply:
Done.